### PR TITLE
BRANCH 4.4: made notes dynamic

### PR DESCRIPTION
**Pull request by remkonoteboom**

### CODE REVIEW

File: src/tactic/ui/container/tab_wdg.py [hunk 1]
```diff
@@ -908,12 +929,27 @@ def get_config_xml(my):
...
+
         my.search_type = None
 
         my.view = my.kwargs.get("view")
         config_xml = my.kwargs.get("config_xml")
         config = my.kwargs.get("config")
 
+        my.save_state = my.kwargs.get("save_state")
+        if my.save_state in [True, 'true']:
+            my.save_state = "main_tab_save_state"
```
Comment (oboreo) [inline src/tactic/ui/container/tab_wdg.py]: can we just call this "main_tab"?

File: src/tactic/ui/container/tab_wdg.py [hunk 2]
```diff
@@ -38,6 +38,11 @@ class TabWdg(BaseRefreshWdg):
              'values': 'true|false',
             'category': 'Display'
         },
+        'save_state': {
+            'description': 'key which is used to save state',
```
Comment (oboreo) [inline src/tactic/ui/container/tab_wdg.py]: examples here would be good like  "save_state|main_tab"

File: src/tactic/ui/widget/discussion_wdg.py [hunk 3]
```diff
@@ -1257,248 +1286,154 @@ def get_display(my):
...
+                var top = bvr.src_el.getParent(".spt_discussion_context_top");
+                if (top.getAttribute("spt_is_loaded") == "true") {
+                    return;
+                }
 
-    def get_context_wdg(my, process, context, show_context_notes=True):
-        ''' this is drawn per process/context group of notes'''
-        div = DivWdg()
-        div.add_class("hand")
+                var class_name = 'tactic.ui.widget.NotesContextWdg';
```
Comment (oboreo) [inline src/tactic/ui/widget/discussion_wdg.py]: Let's rename this as NoteCollectionWdg , keeping the word singular as well.

File: src/tactic/ui/widget/discussion_wdg.py [hunk 4]
```diff
@@ -1257,248 +1286,154 @@ def get_display(my):
...
-                ''' %context
-
-        bvr = {
-                'type': 'click_up',
-                'cbjs_action':  '''
-                %s
-                %s
-                '''%(js_action, swap.get_swap_script())
-                }
+            notes_wdg = NotesContextWdg(
```
Comment (oboreo) [inline src/tactic/ui/widget/discussion_wdg.py]: Let's rename this as NoteCollectionWdg

File: src/tactic/ui/widget/discussion_wdg.py [hunk 5]
```diff
@@ -1538,35 +1473,160 @@ def get_context_wdg(my, process, context, show_context_notes=True):
...
         count_div.add_style("font-size: 1.0em")
         count_div.add_style("font-style: italic")
+        count_div.add_style("margin-left: 3px")
+
+        return div
+
+
+
+__all__.append("NotesContextWdg")
+
```
Comment (oboreo) [inline src/tactic/ui/widget/discussion_wdg.py]: Let's rename this as NoteCollectionWdg, and put this __all__ at the very top

File: src/tactic/ui/widget/discussion_wdg.py [hunk 6]
```diff
@@ -1538,35 +1473,160 @@ def get_context_wdg(my, process, context, show_context_notes=True):
...
+        )
+
+        div.add(widget)
         return div
 
 
 
 
+class NoteWdg(BaseRefreshWdg):
+
```
Comment (oboreo) [inline src/tactic/ui/widget/discussion_wdg.py]: add a comment to it to read something like
'''Display of a single note used by NoteCollectionWdg'''

File: src/tactic/ui/container/tab_wdg.py (hunk 1 above)
Comment (remkonoteboom) [inline src/tactic/ui/container/tab_wdg.py]: Sure ... doesn't really matter what it is.  It's just a look up key.

File: src/tactic/ui/container/tab_wdg.py (hunk 2 above)
Comment (remkonoteboom) [inline src/tactic/ui/container/tab_wdg.py]: Yes, any tab can specify a save state: you just have to specify a unique key to store this under.

File: src/tactic/ui/widget/discussion_wdg.py (hunk 3 above)
Comment (remkonoteboom) [inline src/tactic/ui/widget/discussion_wdg.py]: ok, done

File: src/tactic/ui/widget/discussion_wdg.py (hunk 4 above)
Comment (remkonoteboom) [inline src/tactic/ui/widget/discussion_wdg.py]: done.

File: src/tactic/ui/widget/discussion_wdg.py (hunk 5 above)
Comment (remkonoteboom) [inline src/tactic/ui/widget/discussion_wdg.py]: done

File: src/tactic/ui/widget/discussion_wdg.py (hunk 6 above)
Comment (remkonoteboom) [inline src/tactic/ui/widget/discussion_wdg.py]: done

**Pull request merged**